### ISSUE TITLE
Taming media types

### ISSUE THREAD

**Issue by jonjohnsonjr** [opencontainers/image-spec#791]
Posted this in https://github.com/opencontainers/image-spec/issues/787 originally but noticed it was closed, so I doubt anyone would ever see it.

~~Do we care if OCI media types are valid [RFC6838](https://tools.ietf.org/html/rfc6838#section-4.2) media types? E.g. "zstd" isn't defined in the [Structured Syntax Suffix Registry](https://www.iana.org/assignments/media-type-structured-suffix/media-type-structured-suffix.xhtml).~~ 

>    Media types that make use of a named structured syntax SHOULD use the appropriate registered "+suffix" for that structured syntax when they are registered.  By the same token, media types MUST NOT be given names incorporating suffixes for structured syntaxes they do not actually employ. "+suffix" constructs for as-yet unregistered structured syntaxes SHOULD NOT be used, given the possibility of conflicts with future suffix definitions.

This ties into https://github.com/opencontainers/image-spec/issues/747 as well. I'm not sure if the `+enc` double-suffix is valid (though I mostly skimmed that RFC and might misunderstand the grammar described).

I was thinking about how I'd go about adding support for zstd in various places, and I'm a bit worried that, as we add more dimensions to the media types, we're going to end up with a combinatoric explosion of valid media type strings. As a registry/client author, it's currently easy enough to hard-code these strings and special-case e.g. nondistributable layers or the compression algorithm; however, I can imagine that comparing hard-coded strings will become sub-optimal at some point in the future, and we'd be better off parsing these strings into a structured form.

1. Any thoughts on whether we should try to follow that RFC? FWIW, there's a [procedure](https://tools.ietf.org/html/rfc6838#section-6) for registering a suffix, if anyone wants to take a stab at adding +zstd.
2. Is there blessed code anywhere for parsing OCI media types? Or is everyone just directly comparing [these strings](https://github.com/opencontainers/image-spec/blob/775207bd45b6cb8153ce218cc59351799217451f/specs-go/v1/mediatype.go)?

cc @vbatts @lumjjb @stevvooe

**Comment by ikaneshiro** [comment]
@jonjohnsonjr @vsoch 
The spelling mistake, tar extension and lack of version number on the layer `mediaType` are oversights on my part. I plan on making the `mediaType` name in line with what @SteveLasker has [referenced](https://github.com/SteveLasker/artifacts/blob/artifact-authors-how-to/authoring-artifacts.md#layer-file-format) as it falls under the artifact umbrella.

I'll move the discussion around that to https://github.com/sylabs/singularity/issues/4437 though. Don't want to detract from this more general discussion.

**Comment by vsoch** [comment]
Thank you @ikaneshiro !

**Comment by dmcgowan** [comment]
I put this on the agenda for tomorrow's weekly dev meeting. If we can't get enough people there tomorrow to discuss we can push it out a week too.

**Comment by vbatts** [comment]
i will not be able to make the call, but I feel the discussion is on the
right track. I do like the way Jon laid out an interface for it, but since
order matters, this feels like it'll just be a huge switch statement that
is comparing whole strings.

**Comment by jdolitsky** [comment]
I'd like more clarification somewhere on valid mediatypes (maybe that is the artifacts repo?) .. the Helm mediatypes are using `+tar`/ `+json` as that is the recommendation we had at the time, but now sounding like an oversight https://v3.helm.sh/docs/topics/registries/#where-are-my-charts

**Comment by mikebrow** [comment]
I'd like to see scenarios drawn out for how the +encrypted/encoded layer type suffix would work from build to snapshot for the community at large, before we say yeah let's support a generic media type suffix. Has that been done yet?

It would make more sense to me if it was say +multipart/encrypted with a first portion of the binary providing the necessary detail for decryption and/or +type.encrypted where type indicates which of a plurality of generally supported encryption algorithms for existing and new encryption types possibly specified over here [iana](https://www.iana.org/assignments/media-types/media-types.xhtml) but where we could also add more as needed.

**Comment by lumjjb** [comment]
>I'd like to see scenarios drawn out for how the +encrypted/encoded layer type suffix would work from build to snapshot for the community at large, before we say yeah let's support a generic media type suffix. Has that been done yet?

Yup. @stefanberger, @dmcgowan worked on this together with the [new stream plugin](https://github.com/containerd/containerd/pull/3482) from @crosbymichael. There is also an implementation in `containers/image` library as an [open PR](https://github.com/containers/image/pull/673) - with consumption by skopeo and cri-o.

**Comment by mikebrow** [comment]
> > I'd like to see scenarios drawn out for how the +encrypted/encoded layer type suffix would work from build to snapshot for the community at large, before we say yeah let's support a generic media type suffix. Has that been done yet?
> 
> Yup. @stefanberger, @dmcgowan worked on this together with the [new stream plugin](https://github.com/containerd/containerd/pull/3482) from @crosbymichael. There is also an implementation in `containers/image` library as an [open PR](https://github.com/containers/image/pull/673) - with consumption by skopeo and cri-o.

I see code, but I don't see soup to nuts scenario descriptions. I think I remember seeing some decks where scenarios were was talked about in general but don't see it in the links. Thinking about this from a spec perspective.

I don't see in the code, yet, how you handle all the various encryption types, is there a plain text header in the layer, or some magic number?

**Comment by jonjohnsonjr** [comment]
Posting back here, @estesp kindly took some notes in the OCI call: https://hackmd.io/El8Dd2xrTlCaCG59ns5cwg#Notes1

Some action items from @dmcgowan
1. Add a description to image-spec of the media type format. We should describe existing media types and probably reference the RFC. Perhaps some guidance for clients of how to interpret the media types, but that is covered by the RFC.
2. Update the artifacts repo with explicit guidance for creating new media types that are easily consumable by clients.
3. Come up with a plan for zstd compatibility in 1.1 release

**Comment by jdolitsky** [comment]
@jonjohnsonjr it looks like action item 2 is already in WIP: https://github.com/opencontainers/artifacts/pull/6

There seems to be alot of examples of using `+` there in mediatypes.. we can probably use some comments there on what's appropriate.

**Comment by jzelinskie** [comment]
I think we don't need to require that a compression format be a part of RFC6838, but rather all maintainers should agreed that the format should be added and under what name. This will avoid registry authors from supporting every type of compression under the sun and if there is a format that would be valuable to add, we don't have to wait for anyone else to be able to add it to the standard.

We should keep artifacts discussion on that GitHub repository's issue tracker and scope this discussion purely to application of our spec to containers.

**Comment by jonjohnsonjr** [comment]
Something I just ran into re: nondistributable layers... this mostly involves distribution, but most of the relevant conversation has been happening in this repo.

This PR says that unrecognized media types must be ignored: https://github.com/opencontainers/image-spec/pull/759

Which was the result of this issue to ensure flexibility in media types: https://github.com/opencontainers/distribution-spec/issues/55

So I'm wondering, from a distribution perspective, what it means to _ignore_ something? My first impression was to just completely ignore it by not even validating that the descriptor references a blob that has actually been uploaded, but I think that's probably unhelpful behavior. Nondistributable media types seem like an exception that proves the rule here; i.e. that any non-manifest _should_ be expected to reference an uploaded blob, _unless_ it's a nondistributable media type.

That thought reminded me of this thread, where we're throwing in the towel on nondistributable media types, but I think that's a bad idea. It would be great if we could define an annotation for nondistributability such that non-layer media types can also be outside of the blobstore, e.g. for directly referencing blobs on GitHub instead of within GCR (inspired by discussion with @vbatts).

WDYT @dmcgowan

**Comment by jonjohnsonjr** [comment]
> Do we care if OCI media types are valid RFC6838 media types? E.g. "zstd" isn't defined in the Structured Syntax Suffix Registry.

FWIW [this is no longer true ](https://github.com/facebook/zstd/issues/1846)🎉

**Comment by jonjohnsonjr** [comment]
This is a real problem, not just for me: https://github.com/moby/buildkit/pull/1879

cc @tonistiigi come share in my pain
